### PR TITLE
i#6187 Add invariant check that timestamps never decrease

### PR DESCRIPTION
**Pull request by ivankyluk**

### CODE REVIEW

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -1407,14 +1407,61 @@ check_filter_endpoint()
     return true;
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:1407]: Title should start with "i#NNN: " if there is an associated issue: so "i#6187: Add ...".

Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:1407]: PR description body should include a reference to the issue at the end: either "Issue: #6187" or "Fixes #6187".  Plus generally it should include some text even if it repeats what is in the title.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 2]
```diff
@@ -1407,14 +1407,61 @@ check_filter_endpoint()
...
 }
 
+bool
+check_timestamps_increase_monotonically(void)
+{
+    // Positive test: timestamps increase monotonically.
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:1413]: We've switched from "positive" vs "negative" to "correct" vs "incorrect" in new tests b/c there was some confusion over the former.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 3]
```diff
@@ -1407,14 +1407,61 @@ check_filter_endpoint()
...
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 0),
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:1416]: We've also started using a local constant like `TID` or putting `/*tid=*/, ` to make it easier to understand what this 1 constant is.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 4]
```diff
@@ -604,6 +604,24 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
 
     if (memref.marker.type == TRACE_TYPE_MARKER &&
         memref.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP) {
+#ifdef X86_32
+        // i#5634: Truncated for 32-bit, as documented.
+        // A 32 bit timestamp rolls over every 4294 seconds, so it needs to be
+        // considered when timetsamps are compared. The check assumes two
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:610]: spelling: "timestamps"

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 5]
```diff
@@ -604,6 +604,24 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+        // consecutive timestamps will never be more than 2^31 microseconds (
+        // 2147 seconds) apart.
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:612]: nit: Normally the opening paren would be on the next line: doesn't seem a normal line break point

File: clients/drcachesim/tracer/instru_online.cpp [hunk 6]
```diff
@@ -206,8 +206,10 @@ online_instru_t::refresh_unit_header_timestamp(byte *buf_ptr, uint64 min_timesta
     stamp++; // Skip the tid added by append_unit_header() before the timestamp.
     DR_ASSERT(stamp->type == TRACE_TYPE_MARKER &&
               stamp->size == TRACE_MARKER_TYPE_TIMESTAMP);
-    if (stamp->addr < min_timestamp) {
-        stamp->addr = static_cast<uintptr_t>(min_timestamp);
+    // i#5634: Truncated for 32-bit, as documented.
+    const uintptr_t new_timestamp = static_cast<uintptr_t>(min_timestamp);
+    if (stamp->addr < new_timestamp) {
```
Comment (derekbruening) [inline clients/drcachesim/tracer/instru_online.cpp:211]: Mention this bug fix in the PR description body: this PR contains more than just the invariant check.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1407]: done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1413]: I've changed postive/negative to correct/incorrect.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 4 above)
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp:610]: fixed.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 3 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1416]: I've changed the code to use a constant TID.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 5 above)
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp:612]: fixed.

File: clients/drcachesim/tracer/instru_online.cpp (hunk 6 above)
Comment (ivankyluk) [reply clients/drcachesim/tracer/instru_online.cpp:211]: done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1407]: done.

Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1407]: nit: There should be a colon after the number.

Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1407]: I've updated the title accordingly.

Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:1407]: It looks like this was not actually done?  The body is empty, as seen above and in the actual commit [76b5455](https://github.com/DynamoRIO/dynamorio/commit/76b545554ce3c7e350487ea009cab9fa19596872) which thus ends up with no link to is issue #6187 and no auto-close.  I added a ref there and closed that issue manually.

**Pull request merged**